File: eli_ble_gatt_server/ios/Classes/EliBleGattServerPlugin.swift
import Flutter
import UIKit

public class EliBleGattServerPlugin: NSObject, FlutterPlugin, FlutterStreamHandler {

    private let manager = BleGattServerManager()
    private var eventSink: FlutterEventSink?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance = EliBleGattServerPlugin()

        let methodChannel = FlutterMethodChannel(
            name: "eli_ble_gatt_server",
            binaryMessenger: registrar.messenger()
        )
        registrar.addMethodCallDelegate(instance, channel: methodChannel)

        let eventChannel = FlutterEventChannel(
            name: "eli_ble_gatt_server/events",
            binaryMessenger: registrar.messenger()
        )
        eventChannel.setStreamHandler(instance)

        // El manager emite eventos hacia el sink activo (ya en main thread).
        instance.manager.onEvent = { [weak instance] event in
            instance?.eventSink?(event)
        }
    }

    // MARK: - FlutterPlugin

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "getPlatformVersion":
            result("iOS " + UIDevice.current.systemVersion)

        case "configureServer":
            guard let args = call.arguments as? [String: Any],
                  let serviceUuid = args["serviceUuid"] as? String,
                  let characteristicUuid = args["characteristicUuid"] as? String else {
                result(FlutterError(code: "BAD_ARGS",
                                    message: "serviceUuid and characteristicUuid are required",
                                    details: nil))
                return
            }
            let deviceName = args["deviceName"] as? String
            let payload = args["payload"] as? [String: Any]
            manager.configure(serviceUuid: serviceUuid,
                              characteristicUuid: characteristicUuid,
                              deviceName: deviceName,
                              payload: payload)
            result(true)

        case "startServer":
            manager.start()
            result(true)

        case "stopServer":
            manager.stop()
            result(true)

        case "getServerStatus":
            result(manager.statusMap())

        case "sendMessage":
            guard manager.isActive else {
                result(FlutterError(code: "SERVICE_NOT_RUNNING",
                                    message: "BLE server is not running",
                                    details: nil))
                return
            }
            let args = call.arguments as? [String: Any]
            let message = args?["message"] as? String ?? ""
            manager.sendMessage(message)
            result(true)

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - FlutterStreamHandler

    public func onListen(withArguments arguments: Any?,
                         eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events
        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        self.eventSink = nil
        return nil
    }
}

File: eli_ble_gatt_server/android/src/main/kotlin/com/example/eli_ble_gatt_server/EliBleGattServerPlugin.kt
package com.example.eli_ble_gatt_server

import android.Manifest
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import androidx.core.app.ActivityCompat
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.plugin.common.EventChannel

/** EliBleGattServerPlugin */
class EliBleGattServerPlugin :
    FlutterPlugin,
    MethodCallHandler {

    private lateinit var methodChannel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private lateinit var context: Context

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        context = binding.applicationContext

        // MethodChannel (acciones puntuales)
        methodChannel = MethodChannel(
            binding.binaryMessenger,
            "eli_ble_gatt_server"
        )
        methodChannel.setMethodCallHandler(this)

        // EventChannel (eventos BLE en tiempo real)
        eventChannel = EventChannel(
            binding.binaryMessenger,
            "eli_ble_gatt_server/events"
        )

        eventChannel.setStreamHandler(object : EventChannel.StreamHandler {
            override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
                // Conectamos el Service con Flutter
                EliBleGattServerService.eventSink = events
            }

            override fun onCancel(arguments: Any?) {
                EliBleGattServerService.eventSink = null
            }
        })
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {

            "getPlatformVersion" -> {
                result.success("Android ${android.os.Build.VERSION.RELEASE}")
            }

            "configureServer" -> {
                val serviceUuid = call.argument<String>("serviceUuid")!!
                val characteristicUuid = call.argument<String>("characteristicUuid")!!
                val deviceName = call.argument<String>("deviceName")
                val payload = call.argument<Map<String, Any>>("payload")

                EliBleGattServerService.configure(
                    serviceUuid,
                    characteristicUuid,
                    deviceName,
                    payload
                )

                result.success(true)
            }

            "startServer" -> {
                val intent = Intent(context, EliBleGattServerService::class.java)
                context.startForegroundService(intent)
                result.success(true)
            }

            "stopServer" -> {
                val intent = Intent(context, EliBleGattServerService::class.java)
                context.stopService(intent)
                result.success(true)
            }

            "getServerStatus" -> {
                result.success(EliBleGattServerService.getStatusMap())
            }

            "sendMessage" -> {
                val message = call.argument<String>("message") ?: ""
                val service = EliBleGattServerService.instance
                if (service == null) {
                    result.error("SERVICE_NOT_RUNNING", "BLE server is not running", null)
                    return
                }
                Thread {
                    if (ActivityCompat.checkSelfPermission(
                            context,
                            Manifest.permission.BLUETOOTH_CONNECT
                        ) == PackageManager.PERMISSION_GRANTED
                    ) {
                        service.sendMessageToConnectedDevices(message)
                    }
                }.start()
                result.success(true)
            }

            else -> result.notImplemented()
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel.setMethodCallHandler(null)
        EliBleGattServerService.eventSink = null
    }
}
